pkg/exprparser/functions_test.go

```diff
@@ -0,0 +1,247 @@
+package exprparser
+
+import (
+	"path/filepath"
+	"testing"
+
+	"github.com/nektos/act/pkg/model"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFunctionContains(t *testing.T) {
+	table := []struct {
+		input    string
+		expected interface{}
+		name     string
+	}{
+		{"contains('search', 'item') }}", false, "contains-str-str"},
+		{`cOnTaInS('Hello', 'll') }}`, true, "contains-str-casing"},
+		{`contains('HELLO', 'll') }}`, true, "contains-str-casing"},
+		{`contains('3.141592', 3.14) }}`, true, "contains-str-number"},
+		{`contains(3.141592, '3.14') }}`, true, "contains-number-str"},
+		{`contains(3.141592, 3.14) }}`, true, "contains-number-number"},
+		{`contains(true, 'u') }}`, true, "contains-bool-str"},
+		{`contains(null, '') }}`, true, "contains-null-str"},
+		{`contains(fromJSON('["first","second"]'), 'first') }}`, true, "contains-item"},
+		{`contains(fromJSON('[null,"second"]'), '') }}`, true, "contains-item-null-empty-str"},
+		{`contains(fromJSON('["","second"]'), null) }}`, true, "contains-item-empty-str-null"},
+		// TODO: investigate why this evaluates to false
+		// {`contains(fromJSON('[true,"second"]'), 'true') }}`, true, "contains-item-bool-arr"},
+		// TODO: investigate why this evaluates to false
+		// {`contains(fromJSON('["true","second"]'), true) }}`, true, "contains-item-str-bool"},
```

Comment:
missed?

Reply:
No, these are behaving strange on github and we are not sure what to do with them.
Therefore we left the TODO comment and defered that to a later point.

Reply:
they evaluate to `false` because you compare `true` with `"true"`
```suggestion
		// this looks for string true in array containing boolean true
		// {`contains(fromJSON('[true,"second"]'), 'true') }}`, true, "contains-item-bool-arr"},
		// this looks for boolean true in array containing string true
		// {`contains(fromJSON('["true","second"]'), true) }}`, true, "contains-item-str-bool"},
```

Reply:
can you link me the docs article?

Reply:
Sure: https://docs.github.com/en/actions/learn-github-actions/expressions#functions

Reply:
I think the docs are correct. Since the inserted values are an `Array` in this case. And searched `item` can't be casted to other type otherwise it would introduce a bug.
> Some functions cast values to a string to perform comparisons. GitHub casts data types to a string using these conversions:

![image](https://user-images.githubusercontent.com/31106839/146552725-7cf2bbab-f50e-4d7a-a1ae-676e3aecc0d5.png)

Reply:
@zaubernerd Then we should change the impl to not coerce Array items.

Reply:
@catthehacker see also this workflow, where we tried the above cases on GH actions: https://github.com/ZauberNerd/act-test/runs/4559124896?check_suite_focus=true#step:3:251

Reply:
@KnisterPeter we don't coerce the Array to a string. `contains` accepts either an array or a string as its input.
And the weird thing is that `contains([3.14, 'foo'], '3.14')` and `contains(['3.14', 'bar'], 3.14)` seem to cast values in the array to a string while comparing.

Reply:
@ZauberNerd I'm unsure what you're trying to show me. All cases in that workflow align with what I'm saying and the docs.

Reply:
@zaubernerd we are coercing all Array items to string. That's what we need to change

Reply:
@KnisterPeter although now that I think of it, there might be a bug. This seem to be type casted while it shouldn't
```yaml
[ "${{ contains(fromJSON('[null,"second"]'), '') }}" = "true" ]
[ "${{ contains(fromJSON('["","second"]'), null) }}" = "true" ]
```

Reply:
As I've said before, we are coercing all Array items which we shouldn't do. Our implementation is just wrong.

Reply:
@catthehacker yes, exactly that. GitHub's behaviour seems inconsistent here.
The above linked workflow has these lines in it, some of which indicate that GH does type coercions, but then the `true` case does not which contradicts the observation and the docs:
```
[ "${{ contains(fromJSON('[null,"second"]'), '') }}" = "true" ]
[ "${{ contains(fromJSON('["","second"]'), null) }}" = "true" ]
# TODO: investigate why this evaluates to false
# [ "${{ contains(fromJSON('[true,"second"]'), 'true') }}" = "true" ]
[ "${{ contains(fromJSON('["true","second"]'), 'true') }}" = "true" ]
# TODO: investigate why this evaluates to false
# [ "${{ contains(fromJSON('["true","second"]'), true) }}" = "true" ]
[ "${{ contains(fromJSON('[3.14,"second"]'), '3.14') }}" = "true" ]
[ "${{ contains(fromJSON('[3.14,"second"]'), 3.14) }}" = "true" ]
[ "${{ contains('3.141592', 3.14) }}" = "true" ]
[ "${{ contains(3.141592, '3.14') }}" = "true" ]
[ "${{ contains(3.141592, 3.14) }}" = "true" ]
```

Reply:
Maybe @ChristopherHX can help us out here too :)

Reply:
ref: 
https://github.com/actions/runner/blob/be9632302ceef50bfb36ea998cea9c94c75e5d4d/src/Sdk/DTPipelines/Pipelines/ContextData/JTokenExtensions.cs#L15

Reply:
What happens is if left side is `string`, it compares `string == string`
If left side is an `Array` (https://github.com/actions/runner/blob/be9632302ceef50bfb36ea998cea9c94c75e5d4d/src/Sdk/DTExpressions2/Expressions2/Sdk/Functions/Contains.cs#L26-L29) it does special comparison of objects inside the array (https://github.com/actions/runner/blob/be9632302ceef50bfb36ea998cea9c94c75e5d4d/src/Sdk/DTExpressions2/Expressions2/EvaluationResult.cs#L227-L272)

Reply:
and most important code for you
```csharp
            // Same kind
            if (leftKind == rightKind)
            {
            [...]
            }

            return false;
```
https://github.com/actions/runner/blob/be9632302ceef50bfb36ea998cea9c94c75e5d4d/src/Sdk/DTExpressions2/Expressions2/EvaluationResult.cs#L234

Reply:
ref: https://github.com/actions/runner/blob/be9632302ceef50bfb36ea998cea9c94c75e5d4d/src/Sdk/DTExpressions2/Expressions2/EvaluationResult.cs#L81

Reply:
@catthehacker updated: https://github.com/nektos/act/pull/908/commits/73cf7e35b4765b1133f95dfcb303428997237ab2